libs/shared-utils/src/lib/yup.ts

```diff
@@ -0,0 +1,58 @@
+import { useFormik, FormikConfig, FormikHelpers } from 'formik';
+import * as Yup from 'yup';
+
+type ValidationObjectType<T extends Yup.ObjectSchema<object>> =
+  T extends Yup.ObjectSchema<infer G> ? G : never;
+
+type YupConstructor<T extends object> =
+  (yup: typeof Yup) => Yup.ObjectSchemaDefinition<T>;
+
+type FormikOnSubmit<T extends object, C extends object> =
+  (data: T, helpers: FormikHelpers<T>, context: C) => void;
+
+interface FormikModifiedConfig<T extends object> {
+  /** Should Formik reset the form when new initialValues change */
+  enableReinitialize?: FormikConfig<T>['enableReinitialize'];
+  /** Tells Formik to validate upon mount */
+  validateOnMount?: FormikConfig<T>['validateOnMount'];
+}
+
+function onSubmitFactory<T extends object, C extends object> (
+  validationSchema: Yup.ObjectSchema<T>,
+  initialValues: (context: C) => T,
+  config?: FormikModifiedConfig<T>
+) {
+  return function (onSubmit: FormikOnSubmit<T, C>) {
+    return (context: C) => useFormik<T>({
+      ...(config || {}),
+      validationSchema,
+      initialValues: initialValues(context),
+      onSubmit: (values, helpers) => onSubmit(values, helpers, context)
+    });
+  }
+}
+
+/** ... */
+export function createComponentForm<C extends object> () {
+  return {
+    /** ... */
+    validation<T extends object> (formSchema: YupConstructor<T>) {
+      const validationSchema = Yup.object(formSchema(Yup));
+      type FormType = ValidationObjectType<typeof validationSchema>;
+      
+      return {
+        /** Definition of the initial values using the context of the component  */
+        initialValues: (initialValues: (context: C) => FormType) => ({
+            /** ... */
+            formikConfig: (config?: FormikModifiedConfig<FormType>) => ({
+                /** ... */
+                onSubmit: onSubmitFactory(validationSchema, initialValues, config)
+            }),
+            /** ... */
+            onSubmit: onSubmitFactory(validationSchema, initialValues)
+          })
+      };
+    }
+  }
+}
+
```

Comment:
Another very cool idea here, we definitely needed an abstraction for the sake of consistency of the forms in the project. Very well done 💯

However, I have some concerns here:
* Should createComponentForm utility be in a file named `yup.ts`? Yup is a tool to use for validation but this function does way more than accepting validation. We might need to move it somewhere else.
* Also, with this kind of an API (chaining functions to each other), we lose the sake of configuration being order-ignorant. I mean, when anybody uses this function, they're enforced to use the functions to create a form in the exact order you set. Also, for the forms that don't require any validation or initial values, they will need to be sent in the right order with an empty object, just to make sure onSubmit works. I think this problem can be avoided with a good documentation but again, this can be a very mind and time consuming process for a junior developer, which we have time to time, contributing to these projects. Do you have any ideas how we can tackle this challenge? I'm really keen on discussing this further and make it work because this idea is very very cool!

Reply:
I would rename the file to `form-helper.ts` or so, yeah. The folding of the project it's a topic that I still don't want to open until this PR is ready, but there's a big room for improvement when it comes to organizing the whole repo. Some good practices were already applied in the design system library as you can notice.

The order of the functions serves a strong typing purpose that eliminates the inconsistencies between three elements: `formik`, `yup`, and the initial and referenced values. I found situations in many components where the typing casting was covering a bad definition of the variables, leading to unexpected behaviors with `undefined` or even worse scenarios with an absolute mismatch between the use of variables and configurations. This specific order and currying solve that:

- A friendly and unified API for the creation of the logic of forms.
- The possibility of unit testing the form behavior independently of the UI implementation.
- Eliminates the continuous imports done in all components, leading to clearer and smaller bundles.
- Allows using naming conventions of files for separation of concerns and easier PR reviews in the future.
- Keeps the react components cleaner and allows a lot of reusability that is still not implemented. 
- The junior developer doesn't require to fully understand the logic of the `formik` and `yup` libraries, creating an abstraction that makes it way more comprehensible and easy to use.
- The `validation()` step can be used only for type definition, letting `yup` do the parsing for us without any further config, leading to an empty object return in `initialValues()` but still having, in the end, a return `formik` object that you can use with the design components as they are (while we decouple `formik` from them, with is a terrible thing to see and that, in my opinion, makes that something like a Dropdown select has an API 10 times more complicated than this factory and with zero documentation). Showing juniors how to solve complicated problems with easy solutions like this must be also a priority in a learning path.
- I still have to properly document this in the JSDocs, but since everything is pretty straightforward; it makes the logic of the form way more comprehensive with some embed docs than the wild and wrong compositions that I've seen.